hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/MergeOnReadIncrementalRelation.scala

```diff
@@ -60,11 +60,14 @@ case class MergeOnReadIncrementalRelation(override val sqlContext: SQLContext,
 
   override protected def timeline: HoodieTimeline = {
     if (fullTableScan) {
-      metaClient.getCommitsAndCompactionTimeline
+      filterTimelineForIncrementalQueryIfNeeded(metaClient,
```

Comment:
If using `fullTableScan`, change timeline could also change `queryTimestamp` in `HoodieBaseRelation`, and this could affect fetching latest FileSlices in `listLatestFileSlices`.

I think we can still use `metaClient.getCommitsAndCompactionTimeline` for `fullTableScan`

Reply:
isn't this only confined to `MergeOnReadIncrementalRelation` during incremental read? we use full table scan when we want to extend the timeline to include archived, which should still be consistent with which latest instant being visible? not sure if i missed anything

Reply:
Oh, sorry, this is not only related to `fullTableScan`, but also for normal incremental read. If users manually pass `startTimestamp` and `endTimestamp` to perform incremental read, they might not read all data they needed since some instants might be filtered by non-completed commits. This changes the behavior before.

I think in most cases, we concern streaming jobs(like `HoodieStreamingSource`) miss non-complete commits, but for incremental jobs which the user specifies `endTimestamp`, do we need to also keep strict behavior here?

Reply:
I have similiar concern, for batch incremental queries with explicit start and end timestamp, there is no need to skip any completed commiits. The invoker should handle the 'hollow' instants (inflight in current time range while completes later) by themselves.

And for streaming source, we could checkpoint both the max completion time(transition time) and the max version id(instant time), these 2 timestamp could help to find out the 'hollow' instants and fix the reader automatically. The streaming reader fix makes more sense generally to users.

Reply:
from offline discussion with @danny0405 and @codope : the behavior should be, for batch incremental query when we detect such in-between inflight commits, we throw an exception to alert user, regardless of user providing end timestamp or not. For streaming read, we recommend users to turn on use transition time if multi-writer is enabled. Changes to include transition time in the streaming checkpoint will be deferred for now.

Reply:
If user doesn't provide the end timestamp, we still throw an exception if finding hollow instants? How abt setting the firstIncompleteCommit as end timestamp?

Throw exceptions disrupts our concurrency control to some extent(In the subconscious of many users, reader side won't be affected if using multi writers for hudi), but for now it looks we don't have other easy workarounds. I think we can recommend users to turn on use-transition-time if multi-writer is enabled both for incremental query & streaming read.

Reply:
throwing an exception is a way to alert users so that they do not blindly query again with the latest timestamp under the impression that all data up to date was return, as we do not return cursor from in the incremental query result. We only do this when not using transition time. 

> we can recommend users to turn on use-transition-time if multi-writer is enabled both for incremental query & streaming read

that's the intention. we can recommend users for configuring it.